rust/types/src/execution/operator.rs

```diff
@@ -637,3 +661,861 @@ impl TryFrom<KnnBatchResult> for chroma_proto::KnnBatchResult {
+                    .collect::<Vec<_>>();
+                sorted_bits.sort_unstable();
+                sorted_bits.hash(state);
+            }
+        }
+    }
+}
+
+impl TryFrom<chroma_proto::QueryVector> for QueryVector {
+    type Error = QueryConversionError;
+
+    fn try_from(value: chroma_proto::QueryVector) -> Result<Self, Self::Error> {
+        let vector = value.vector.ok_or(QueryConversionError::field("vector"))?;
+        match vector {
+            chroma_proto::query_vector::Vector::Dense(dense) => {
+                Ok(QueryVector::Dense(dense.try_into().map(|(v, _)| v)?))
+            }
+            chroma_proto::query_vector::Vector::Sparse(sparse) => {
+                Ok(QueryVector::Sparse(sparse.into()))
+            }
+        }
+    }
+}
+
+impl TryFrom<QueryVector> for chroma_proto::QueryVector {
+    type Error = QueryConversionError;
+
+    fn try_from(value: QueryVector) -> Result<Self, Self::Error> {
+        match value {
+            QueryVector::Dense(vec) => {
+                let dim = vec.len();
+                Ok(chroma_proto::QueryVector {
+                    vector: Some(chroma_proto::query_vector::Vector::Dense(
+                        chroma_proto::Vector::try_from((vec, ScalarEncoding::FLOAT32, dim))?,
+                    )),
+                })
+            }
+            QueryVector::Sparse(sparse) => Ok(chroma_proto::QueryVector {
+                vector: Some(chroma_proto::query_vector::Vector::Sparse(sparse.into())),
+            }),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct KnnQuery {
+    pub embedding: QueryVector,
+    pub key: String,
+    pub limit: u32,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum Rank {
+    #[serde(rename = "$abs")]
+    Absolute(Box<Rank>),
+    #[serde(rename = "$div")]
+    Division { left: Box<Rank>, right: Box<Rank> },
+    #[serde(rename = "$exp")]
+    Exponentiation(Box<Rank>),
+    #[serde(rename = "$knn")]
+    Knn {
+        embedding: QueryVector,
+        #[serde(default = "Rank::default_knn_key")]
+        key: String,
+        #[serde(default = "Rank::default_knn_limit")]
+        limit: u32,
+        #[serde(default)]
+        default: Option<f32>,
+        #[serde(default)]
+        ordinal: bool,
+    },
+    #[serde(rename = "$log")]
+    Logarithm(Box<Rank>),
+    #[serde(rename = "$max")]
+    Maximum(Vec<Rank>),
+    #[serde(rename = "$min")]
+    Minimum(Vec<Rank>),
+    #[serde(rename = "$mul")]
+    Multiplication(Vec<Rank>),
+    #[serde(rename = "$sub")]
+    Subtraction { left: Box<Rank>, right: Box<Rank> },
+    #[serde(rename = "$sum")]
+    Summation(Vec<Rank>),
+    #[serde(rename = "$val")]
+    Value(f32),
+}
+
+impl Rank {
+    pub fn default_knn_key() -> String {
+        "#embedding".to_string()
+    }
+
+    pub fn default_knn_limit() -> u32 {
+        128
+    }
+
+    pub fn knn_queries(&self) -> HashSet<KnnQuery> {
```

Comment:
When we build the Rank expression, lets instead add a offset/ordinal to the KNN, and use that as opposed to hashing. Can TODO in a follow up